File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/P3/CommunityFeedDTOs.swift
//
//  CommunityFeedDTOs.swift
//  Pantopus
//
//  DTOs for the Phase-3 community mail feed in
//  `backend/routes/mailboxV2Phase3.js` (mounted at
//  `/api/mailbox/v2/p3` — `backend/app.js:317`).
//
//  The feed rows are raw `CommunityMailItem` table rows (`select('*')`,
//  route line 575) enriched server-side with a `reactions` roll-up and
//  the caller's own `user_reactions`. Only the columns the native screen
//  renders are modelled.
//

import Foundation

/// One aggregated reaction bucket on a community item —
/// `backend/routes/mailboxV2Phase3.js:609`.
public struct CommunityReactionCountDTO: Decodable, Sendable, Hashable {
    /// `acknowledged / will_attend / concerned / thumbs_up`
    /// (validator at `backend/routes/mailboxV2Phase3.js:51`).
    public let reactionType: String
    public let count: Int

    private enum CodingKeys: String, CodingKey {
        case count
        case reactionType = "reaction_type"
    }

    public init(reactionType: String, count: Int) {
        self.reactionType = reactionType
        self.count = count
    }
}

/// A published neighborhood / civic item.
/// Route: `backend/routes/mailboxV2Phase3.js:565`.
public struct CommunityFeedItemDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let mailId: String?
    public let homeId: String?
    /// `civic_notice / neighborhood_event / local_business /
    /// building_announcement` (`categoryCommunityType`, route line 812).
    public let communityType: String?
    /// `building / neighborhood / city`.
    public let publishedTo: String?
    public let title: String?
    public let body: String?
    public let senderDisplay: String?
    public let senderTrust: String?
    public let category: String?
    public let verifiedSender: Bool?
    public let eventDate: String?
    public let rsvpDeadline: String?
    public let views: Int?
    public let neighborsReceived: Int?
    public let rsvpCount: Int?
    public let createdAt: String?
    public let reactions: [CommunityReactionCountDTO]?
    public let userReactions: [String]?

    private enum CodingKeys: String, CodingKey {
        case id, title, body, category, views, reactions
        case mailId = "mail_id"
        case homeId = "home_id"
        case communityType = "community_type"
        case publishedTo = "published_to"
        case senderDisplay = "sender_display"
        case senderTrust = "sender_trust"
        case verifiedSender = "verified_sender"
        case eventDate = "event_date"
        case rsvpDeadline = "rsvp_deadline"
        case neighborsReceived = "neighbors_received"
        case rsvpCount = "rsvp_count"
        case createdAt = "created_at"
        case userReactions = "user_reactions"
    }

    public init(
        id: String,
        mailId: String? = nil,
        homeId: String? = nil,
        communityType: String? = nil,
        publishedTo: String? = nil,
        title: String? = nil,
        body: String? = nil,
        senderDisplay: String? = nil,
        senderTrust: String? = nil,
        category: String? = nil,
        verifiedSender: Bool? = nil,
        eventDate: String? = nil,
        rsvpDeadline: String? = nil,
        views: Int? = nil,
        neighborsReceived: Int? = nil,
        rsvpCount: Int? = nil,
        createdAt: String? = nil,
        reactions: [CommunityReactionCountDTO]? = nil,
        userReactions: [String]? = nil
    ) {
        self.id = id
        self.mailId = mailId
        self.homeId = homeId
        self.communityType = communityType
        self.publishedTo = publishedTo
        self.title = title
        self.body = body
        self.senderDisplay = senderDisplay
        self.senderTrust = senderTrust
        self.category = category
        self.verifiedSender = verifiedSender
        self.eventDate = eventDate
        self.rsvpDeadline = rsvpDeadline
        self.views = views
        self.neighborsReceived = neighborsReceived
        self.rsvpCount = rsvpCount
        self.createdAt = createdAt
        self.reactions = reactions
        self.userReactions = userReactions
    }
}

/// Envelope for `GET /api/mailbox/v2/p3/community/feed` —
/// `{ items, total }` (route line 617).
public struct CommunityFeedResponse: Decodable, Sendable, Hashable {
    public let items: [CommunityFeedItemDTO]
    public let total: Int?
}

/// Wire body for `POST /api/mailbox/v2/p3/community/react` — validator at
/// `backend/routes/mailboxV2Phase3.js:51`.
public struct CommunityReactRequest: Encodable, Sendable {
    public let communityItemId: String
    public let reactionType: String

    public init(communityItemId: String, reactionType: String) {
        self.communityItemId = communityItemId
        self.reactionType = reactionType
    }
}

/// Envelope for `POST /api/mailbox/v2/p3/community/react` —
/// `{ message, reactions }` (route line 738). The reaction rows are the
/// authoritative post-toggle counts.
public struct CommunityReactResponse: Decodable, Sendable, Hashable {
    public let message: String?
    public let reactions: [CommunityReactionCountDTO]?
}

/// Wire body for `POST /api/mailbox/v2/p3/community/flag` — validator at
/// `backend/routes/mailboxV2Phase3.js:60`.
public struct CommunityFlagRequest: Encodable, Sendable {
    public let communityItemId: String

    public init(communityItemId: String) {
        self.communityItemId = communityItemId
    }
}

/// Envelope for `POST /api/mailbox/v2/p3/community/flag` —
/// `{ message: 'Item flagged for review' }` (route line 805).
public struct CommunityFlagResponse: Decodable, Sendable, Hashable {
    public let message: String?
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/p3/CommunityFeedDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.mailbox.p3

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * DTOs for the Phase-3 community mail feed in
 * `backend/routes/mailboxV2Phase3.js` (mounted at `api/mailbox/v2/p3` —
 * `backend/app.js:317`).
 *
 * Feed rows are raw `CommunityMailItem` table rows enriched server-side
 * with a `reactions` roll-up and the caller's own `user_reactions`. Only
 * the columns the native screen renders are modelled.
 */

/** One aggregated reaction bucket — route `backend/routes/mailboxV2Phase3.js:609`. */
@JsonClass(generateAdapter = true)
data class CommunityReactionCountDto(
    @Json(name = "reaction_type") val reactionType: String,
    @Json(name = "count") val count: Int,
)

/** A published neighborhood / civic item — route `backend/routes/mailboxV2Phase3.js:565`. */
@JsonClass(generateAdapter = true)
data class CommunityFeedItemDto(
    @Json(name = "id") val id: String,
    @Json(name = "mail_id") val mailId: String? = null,
    @Json(name = "home_id") val homeId: String? = null,
    /**
     * `civic_notice` / `neighborhood_event` / `local_business` /
     * `building_announcement` — `categoryCommunityType`,
     * `backend/routes/mailboxV2Phase3.js:812`.
     */
    @Json(name = "community_type") val communityType: String? = null,
    @Json(name = "published_to") val publishedTo: String? = null,
    @Json(name = "title") val title: String? = null,
    @Json(name = "body") val body: String? = null,
    @Json(name = "sender_display") val senderDisplay: String? = null,
    @Json(name = "sender_trust") val senderTrust: String? = null,
    @Json(name = "category") val category: String? = null,
    @Json(name = "verified_sender") val verifiedSender: Boolean? = null,
    @Json(name = "event_date") val eventDate: String? = null,
    @Json(name = "rsvp_deadline") val rsvpDeadline: String? = null,
    @Json(name = "views") val views: Int? = null,
    @Json(name = "neighbors_received") val neighborsReceived: Int? = null,
    @Json(name = "rsvp_count") val rsvpCount: Int? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    @Json(name = "reactions") val reactions: List<CommunityReactionCountDto>? = null,
    @Json(name = "user_reactions") val userReactions: List<String>? = null,
)

/** `GET api/mailbox/v2/p3/community/feed` envelope — route line 617. */
@JsonClass(generateAdapter = true)
data class CommunityFeedResponse(
    @Json(name = "items") val items: List<CommunityFeedItemDto> = emptyList(),
    @Json(name = "total") val total: Int? = null,
)

/**
 * Wire body for `POST api/mailbox/v2/p3/community/react` — validator at
 * `backend/routes/mailboxV2Phase3.js:51`. `reactionType` is one of
 * `acknowledged` / `will_attend` / `concerned` / `thumbs_up`.
 */
@JsonClass(generateAdapter = true)
data class CommunityReactRequest(
    @Json(name = "communityItemId") val communityItemId: String,
    @Json(name = "reactionType") val reactionType: String,
)

/** `POST api/mailbox/v2/p3/community/react` envelope — route line 738. */
@JsonClass(generateAdapter = true)
data class CommunityReactResponse(
    @Json(name = "message") val message: String? = null,
    @Json(name = "reactions") val reactions: List<CommunityReactionCountDto>? = null,
)

/**
 * Wire body for `POST api/mailbox/v2/p3/community/flag` — validator at
 * `backend/routes/mailboxV2Phase3.js:60`.
 */
@JsonClass(generateAdapter = true)
data class CommunityFlagRequest(
    @Json(name = "communityItemId") val communityItemId: String,
)

/** `POST api/mailbox/v2/p3/community/flag` envelope — route line 805. */
@JsonClass(generateAdapter = true)
data class CommunityFlagResponse(
    @Json(name = "message") val message: String? = null,
)
